File: elixir/bob/bob.exs
defmodule Bob do
  def hey(input) do
   cond do
       is_silence(input) -> "Fine. Be that way!"
       is_shouting(input) or is_russian(input) -> "Whoa, chill out!"
       is_question(input)  -> "Sure."
       true -> "Whatever."
   end
  end

  defp is_shouting(input) do
    if String.ends_with?(input, "!") do
        String.match?(input, ~r/^[ !A-Z0-9,%^@#$(*^]*$/)
    else
        String.match?(input, ~r/^[ A-Z]*$/)
    end

  end

  defp is_russian(input) do
    String.match?(input, ~r/^[УХОДИ]*$/)
  end

  defp is_question(input) do
    String.ends_with?(input, "?")
  end

  defp is_silence(input) do
    String.match?(input, ~r/^ *$/)
  end

end

File: haskell/bob/src/Bob.hs
module Bob (responseFor) where

responseFor :: String -> String
responseFor xs
  | silent xs == ""  = "Fine. Be that way!"
  | shoutQuestion xs = "Calm down, I know what I'm doing!"
  | shout xs  = "Whoa, chill out!"
  | question xs = "Sure."
  | otherwise = "Whatever."

silent :: String -> String
silent [] = ""
silent xs = onlyCharacters xs ++ [x | x <- xs, elem x (['0'..'9'] ++ [':','?'])]

shout :: String -> Bool
shout xs = (onlyCharacters xs) /= "" && length([x | x <- onlyCharacters xs, elem x (['A'..'Z'] )] ) == length (onlyCharacters xs)

shoutQuestion :: String -> Bool
shoutQuestion x = shout x && question x

question :: String -> Bool
question x = last x == '?'

onlyCharacters :: String -> String
onlyCharacters xs = [x | x <- xs, elem x  (['A'..'Z'] ++ ['a'..'z'])]

